support for ttl nodes and container nodes

einxie opened the pull request:
Fixes #11

pmazzini commented:
Thanks for the PR!

pmazzini commented:
Test is failing:

```
--- FAIL: TestCreateTTL (1.05s)
##[error]    cluster_test.go:15: [ZKERR] ZooKeeper JMX enabled by default
##[error]    cluster_test.go:15: [ZKERR] Using config: /tmp/gozk460099447/srv1/zoo.cfg
##[error]    zk_test.go:118: Create returned error: unknown error: -6
=== RUN   TestCreateContainer
```

pmazzini commented:
I think that by default TTL support is disabled on the server and needs to be enabled in the integration test.

einxie commented:
> Test is failing:
> 
> ```
> --- FAIL: TestCreateTTL (1.05s)
> ##[error]    cluster_test.go:15: [ZKERR] ZooKeeper JMX enabled by default
> ##[error]    cluster_test.go:15: [ZKERR] Using config: /tmp/gozk460099447/srv1/zoo.cfg
> ##[error]    zk_test.go:118: Create returned error: unknown error: -6
> === RUN   TestCreateContainer
> ```

Perhaps there are two points.
1. Zk's version need to be bigger than 3.5.4(mine is 3.6.0)
2. TTL nodes is disabled by dafault, and need to be enabled by adding two config options as below in zkServer.sh before starting zk server.
-Dzookeeper.extendedTypesEnabled=true -Dzookeeper.emulate353TTLNodes=true
![zkServer](https://user-images.githubusercontent.com/3304342/83364711-cebbd800-a3d5-11ea-8a1d-b67e97aa9fbc.png)

The config infos are from ZooKeeper Administrator's Guide as below:
https://zookeeper.apache.org/doc/r3.5.4-beta/zookeeperAdmin.html

pmazzini commented:
> Zk's version need to be bigger than 3.5.4(mine is 3.6.0)

Yes, I am bumping the versions in #24.

> TTL nodes is disabled by dafault, and need to be enabled by adding two config options as below in zkServer.sh before starting zk server.

Yes, maybe we should enable it here: https://github.com/go-zookeeper/zk/blob/master/server_java_test.go#L51. I can take a look.

yarikk commented on conn.go:
```diff
@@ -1055,6 +1055,33 @@ func (c *Conn) Create(path string, data []byte, flags int32, acl []ACL) (string,
+	if err := validatePath(path, flags&FlagSequence == FlagSequence); err != nil {
+		return "", err
+	}
+	if flags != FlagWithTTL && flags != FlagSequenceWithTTL {
```
shouldn't this be tested as a bites, e.g. `flags&FlagWithTTL != FlagWithTTL`?

einxie replied:
At first I think bit 0 is for Ephemeral，bit 1 is for Sequence，and bit 2 is for TTL，but the value 4(bit 2) is FlagContainer, so I was confused, and listed all the flag values. For example, if I use bit 2 for FlagTTL, when I create container node, I have to use FlagTTL .

einxie replied:
In other way, container node will be deleted in some condition, so it can alse be thinked as ttl. Perhaps bit 2 for FlagTTL is well. I'll modify it.

einxie replied:
> @einxie could you follow up on the above comments please

You're right. Bit 2 for FlagTTL is well. I have modifed it.

yarikk commented on zk_test.go:
```diff
@@ -97,6 +97,84 @@ func TestCreate(t *testing.T) {
+	if err := zk.Delete(path, -1); err != nil && err != ErrNoNode {
+		t.Fatalf("Delete returned error: %+v", err)
+	}
+	if p, err := zk.CreateTTL(path, []byte{1, 2, 3, 4}, FlagWithTTL, WorldACL(PermAll), 60*1000); err != nil {
```
`FlagWithTTL | FlagSequence` to verify the combination working

einxie replied:
Just as 24's answer

pmazzini commented:
@einxie Can you rebase this on #24 and #25? Thanks

pmazzini commented:
`TestCreateTTL` is now failing with:

```
zk_test.go:136: Create returned different path '/gozk-test-ttl0000000001' != '/gozk-test-ttl'
```

einxie commented:
> @einxie could you follow up on the above comments please



> `TestCreateTTL` is now failing with:
> 
> ```
> zk_test.go:136: Create returned different path '/gozk-test-ttl0000000001' != '/gozk-test-ttl'
> ```

The test case has check error and I have modified it.

> `TestCreateTTL` is now failing with:
> 
> ```
> zk_test.go:136: Create returned different path '/gozk-test-ttl0000000001' != '/gozk-test-ttl'
> ```

The check item of SEQUENTIAL TTL node has problem and I have modified it.

pmazzini commented:
`TestCreateTTL` is now failing with:

```
zk_test.go:139: Get returned error: zk: node does not exist
```

I think that the test checks a different path to the one returned by `CreateTTL`.

pmazzini commented on constants.go:
```diff
@@ -27,6 +27,8 @@ const (
 	opCheck        = 13
 	opMulti        = 14
 	opReconfig     = 16
+	opCreateContainer	= 19
```
NIT: this can be better aligned. can you try gofmt?

pmazzini commented on constants.go:
```diff
@@ -72,6 +74,7 @@ const (
 const (
 	FlagEphemeral = 1
 	FlagSequence  = 2
+	FlagTTL = 4
```
NIT: this can be better aligned. can you try gofmt?

pmazzini commented on constants.go:
```diff
@@ -193,6 +196,8 @@ var (
 	opNames       = map[int32]string{
 		opNotify:       "notify",
 		opCreate:       "create",
+		opCreateContainer:  "createContainer",
```
NIT: this can be better aligned. can you try gofmt?

einxie commented:
> `TestCreateTTL` is now failing with:
> 
> ```
> zk_test.go:139: Get returned error: zk: node does not exist
> ```
> 
> I think that the test checks a different path to the one returned by `CreateTTL`
Sorry to make the mistake. Have revised it. Gofmt is also used for the changed files.

pmazzini commented:
Aside from the code coverage, this lgtm. Thanks

einxie commented:
> Aside from the code coverage, this lgtm. Thanks

Thanks. Maybe there are some problem in multi command. I delete multi's support for ttl and container node in this version. Others are not changed.

pmazzini commented:
```
##[error]    TestRequestStructForOp: structs_test.go:26: No struct for op createTTL
##[error]    TestRequestStructForOp: structs_test.go:26: No struct for op createContainer
```

einxie commented:
> ```
> ##[error]    TestRequestStructForOp: structs_test.go:26: No struct for op createTTL
> ##[error]    TestRequestStructForOp: structs_test.go:26: No struct for op createContainer
> ```

Have revised the mistake. Add some testcase for coverage.

jeffbean commented on conn.go:
```diff
@@ -1055,6 +1055,33 @@ func (c *Conn) Create(path string, data []byte, flags int32, acl []ACL) (string,
+}
+
+// ttl: ms
+func (c *Conn) CreateTTL(path string, data []byte, flags int32, acl []ACL, ttl int64) (string, error) {
```
should our Public API in Go use time.Duration for TTL? it's a little awkward to take the int64 as input here.

einxie replied:
Yes, duration is better. I have revised it.

jeffbean commented on structs.go:
```diff
@@ -165,6 +165,16 @@ type CreateRequest struct {
+	Data  []byte
+	Acl   []ACL
+	Flags int32
+	Ttl   int64
```
here it would now be important to comment on the unit of time. Especially since this is an exported type.

einxie replied:
ok

jeffbean commented on conn.go:
```diff
@@ -1055,6 +1055,32 @@ func (c *Conn) Create(path string, data []byte, flags int32, acl []ACL) (string,
+	}
+
+	res := &createResponse{}
+	_, err := c.request(opCreateTTL, &CreateTTLRequest{path, data, acl, flags, ttl.Nanoseconds() / 1e6}, res, nil)
```
you should be able to here do
```
ttl * time.Millisecond
```

we should not need to do any math on this unit to get the underlying int64 that is backing the time.Duration.

einxie replied:
Here I think I can modify to “int64(ttl / time.Millisecond)”

pmazzini commented on conn.go:
```diff
@@ -1055,6 +1055,32 @@ func (c *Conn) Create(path string, data []byte, flags int32, acl []ACL) (string,
+	}
+
+	res := &createResponse{}
+	_, err := c.request(opCreateTTL, &CreateTTLRequest{path, data, acl, flags, int64(ttl / time.Millisecond)}, res, nil)
```
I think `int64(ttl / time.Millisecond)` can be replaced with `ttl.Milliseconds()`

einxie commented:
2020-06-10T09:47:09.2912978Z ##[error]./conn.go:1080:80: ttl.Milliseconds undefined (type time.Duration has no field or method Milliseconds)

There are errors in old version of go1.12.x

pmazzini commented:
>There are errors in old version of go1.12.x

Where are you seeing that? We are supposed to be using go1.13 and go.14:
https://github.com/go-zookeeper/zk/blob/master/.github/workflows/unittest.yaml#L10

einxie commented:
> > There are errors in old version of go1.12.x
> 
> Where are you seeing that? We are supposed to be using go1.13 and go.14:
> https://github.com/go-zookeeper/zk/blob/master/.github/workflows/unittest.yaml#L10

In the checks error log of last version. 
unittest (3.4.14, 1.12.x)/6_Test code.txt
2020-06-10T09:47:08.9103638Z ##[group]Run make test
2020-06-10T09:47:08.9103850Z [36;1mmake test[0m
2020-06-10T09:47:08.9143207Z shell: /bin/bash -e {0}
2020-06-10T09:47:08.9143350Z env:
2020-06-10T09:47:08.9143485Z   GOROOT: /opt/hostedtoolcache/go/1.12.17/x64
2020-06-10T09:47:08.9143620Z   JAVA_HOME: /opt/hostedtoolcache/jdk/9.0.7/x64
2020-06-10T09:47:08.9143756Z   JAVA_HOME_9.0.7_x64: /opt/hostedtoolcache/jdk/9.0.7/x64
2020-06-10T09:47:08.9143929Z ##[endgroup]
2020-06-10T09:47:09.0638835Z go build ./...
2020-06-10T09:47:09.2905627Z # github.com/go-zookeeper/zk
2020-06-10T09:47:09.2912978Z ##[error]./conn.go:1080:80: ttl.Milliseconds undefined (type time.Duration has no field or method Milliseconds)
2020-06-10T09:47:09.2927726Z note: module requires Go 1.13
2020-06-10T09:47:09.6020451Z make: *** [build] Error 2
2020-06-10T09:47:09.6031288Z ##[error]Makefile:38: recipe for target 'build' failed
2020-06-10T09:47:09.6034558Z ##[error]Process completed with exit code 2.

einxie commented:
> > There are errors in old version of go1.12.x
> 
> Where are you seeing that? We are supposed to be using go1.13 and go.14:
> https://github.com/go-zookeeper/zk/blob/master/.github/workflows/unittest.yaml#L10

I retried again and this version of check seems well. Maybe the version checked at that time is error.

pmazzini commented:
thanks!
